src/NexusMods.App.UI/RightContent/Downloads/ViewModels/DownloadTaskViewModel.cs

```diff
@@ -0,0 +1,124 @@
+using NexusMods.DataModel.RateLimiting;
+using NexusMods.Networking.Downloaders.Interfaces;
+using NexusMods.Networking.Downloaders.Interfaces.Traits;
+using ReactiveUI;
+
+namespace NexusMods.App.UI.RightContent.Downloads.ViewModels;
+
+public class DownloadTaskViewModel : AViewModel<IDownloadTaskViewModel>, IDownloadTaskViewModel
+{
+    private readonly IDownloadTask _task;
+    
+    public DownloadTaskViewModel(IDownloadTask task, bool initPreviousStates = true)
+    {
+        _task = task;
+        
+        // Initialize the previous states
+        if (!initPreviousStates)
+            return;
+        
+        _previousName = Name;
+        _previousVersion = Version;
+        _previousGame = Game;
+        _previousStatus = Status;
+        _previousDownloadedBytes = DownloadedBytes;
+        _previousSizeBytes = SizeBytes;
+        _previousThroughput = Throughput;
+    }
+
+    public string Name => _task.FriendlyName;
+    public string Version 
+    {
+        get
+        {
+             if (_task is IHaveDownloadVersion version)
+                 return version.Version;
+
+             return "Unknown";
+        }
+    }
+    
+    public string Game 
+    {
+        get
+        {
+            if (_task is IHaveGameName name)
+                return name.GameName;
+
+            return "Unknown";
+        }
+    }
+
+    public DownloadTaskStatus Status => _task.Status;
+
+    public long DownloadedBytes => (long)_task.DownloadJobs.GetTotalCompletion().Value;
+    
+    public long SizeBytes 
+    {
+        get
+        {
+            if (_task is IHaveFileSize size)
+                return size.SizeBytes;
+
+            return 0;
+        }
+    }
+    
+    public long Throughput => (long)_task.DownloadJobs.GetTotalThroughput(DateTimeProvider.Instance).Value;
+    public void Cancel() => _task.Cancel();
+
+    // Polling implementation, for bridging the gap between a non-INotifyPropertyChanged implementation and 
+    // live-updating ViewModel.
+    private string _previousName = string.Empty;
+    private string _previousVersion = string.Empty;
+    private string _previousGame = string.Empty;
+    private DownloadTaskStatus _previousStatus = DownloadTaskStatus.Idle;
+    private long _previousDownloadedBytes = 0;
+    private long _previousSizeBytes = 0;
+    private long _previousThroughput = 0;
+    
+    public void Poll()
```

Comment:
Not sure that we need this we should be able to do this via `[Reactive]` or observable chains

Reply:
I'm not a fan of it either, but there are specific reasons I wrote it in this way. 

1. Some metrics (e.g. throughput/download speed); inherently require polling.

Download speed is defined as the amount of bytes transferred within the last period of <x> time; and must be polled. In this case, we need some sort of poll loop anyway. 

While in theory we could technically add polling to the HTTP Downloaders themselves to calculate this info, you would still have to poll inside the HTTP Downloaders themselves. In that case, we'd end up having to write more code because both HTTP downloaders would need this mechanism for parity.

2. Some third party APIs don't expose the data we need.

Not 100% sure if we'll run into this, but I've also been burned in the past with implementing downloads from certain services, where the necessary APIs didn't expose events to obtain the data we need (e.g. Updates/Events on Download Progress). 

One notorious case in the past where I've been burned really hard was NuGet, abstracting away access to even the HTTP/Source URL from which the contents are being fetched; and not exposing any info at all. This was a bit problematic... as I couldn't handle the download myself either.

I've used self-hosted NuGet as a form of mod distribution in the past; and I had to use reflection in production to access an internal field to get HTTP URL :sweat_smile:. Not the proudest thing, but I had control over the dependency version so it was my only option short of reimplementing the client myself (download source was deliberately abstracted, so no possible PR).

----------

In any case, this was mostly a code saving measure, a polling mechanism of some sort is necessary anyway, (at least for calculating download speed). 

I figured a single polling loop would have been better than risking having multiple independent polling loops down the road (e.g. in HTTP downloaders, or specific service download implementation).

Reply:
I'd rather this information be in some sort of behind the scenes downloader service, otherwise people will lose the download throughput history everytime they navigate away from a view model (and the VM is destroyed)
 
 >>Some third party APIs don't expose the data we need.
 
 Not sure how this applies to the task at hand, we always have access to the download rate and although we may not have access to the full download size, those sorts of HTTP services are fairly rare.

Reply:
> I'd rather this information be in some sort of behind the scenes downloader service, otherwise people will lose the download throughput history everytime they navigate away from a view model (and the VM is destroyed)

Good point, is it fine if I handle this when we get around to adding the graph to the view?
At that point, keeping this history will be necessary. 

Incremental steps; don't want to hold this PR back longer than it needs to be. 

> those sorts of HTTP services are fairly rare.

Totally agree, but I had to include the example; as it was something I've ran into in the past. While extremely unlikely, and quite honestly I doubt we'll run into it, it is possible.

Writing the code this way was once again, just a cost saving measure as for throughput (etc.), we have to poll anyway.

Reply:
If the tests pass at this point I'll merge this. In the future I think we should probably stop work on a ticket like this and deal with it in smaller PRs, as that will help us avoid situations where we're forced to either do good code reivews or not have something merged for awhile. That and maybe I should put a limit on the number of files someone can edit in a single PR :P

Reply:
Yeah; honestly; this one spun out of control :P

:sweat_smile: 
gosh...